[MLIR][DISC] pattern conversion from tf2mhlo: ConvertUnpackOpDynamic, ConvertSignOpDynamic, ConvertSigmoidGradOpDynamic

azazhu opened the pull request:
We are porting our MLIR-based dynamic shape compiler to tf community (From OP def, Patttern, to Optimization pass, etc).
This is the 5th PR about tf2mhlo pattern conversion, which including ConvertUnpackOpDynamic, ConvertSignOpDynamic, ConvertSigmoidGradOpDynamic.
The rest pattern conversions we will add:
- ConvertSqueezeOpxxx
- ConvertStridedSliceOpxxx
- ConvertPrintOp

azazhu committed:
[MLIR][DISC] pattern conversion from tf2mhlo: ConvertUnpackOpDynamic, ConvertSignOpDynamic, ConvertSigmoidGradOpDynamic

joker-eph commented on tensorflow/compiler/mlir/xla/transforms/legalize_tf.cc:
```diff
@@ -5088,6 +5088,163 @@ class ConvertUnpackOp : public OpRewritePattern<TF::UnpackOp> {
+                                     PatternRewriter& rewriter) const override {
+    auto loc = op.getLoc();
+    auto y = op.y();
+    auto dy = op.dy();
```
Can you replace auto with the type when it isn't obvious from the immediate context and does not improve readability?

azazhu replied:
Done.

joker-eph commented on tensorflow/compiler/mlir/xla/transforms/legalize_tf.cc:
```diff
@@ -5088,6 +5088,163 @@ class ConvertUnpackOp : public OpRewritePattern<TF::UnpackOp> {
+    auto dy = op.dy();
+    auto tp_y = y.getType().dyn_cast<RankedTensorType>();
+    auto tp_dy = dy.getType().dyn_cast<RankedTensorType>();
+    if (!tp_y || !tp_dy || !tp_y.hasRank() || !tp_dy.hasRank())
```
```suggestion
    if (!tp_y || !tp_dy)
```

azazhu replied:
Fixed.

azazhu committed:
[MLIR][DISC] refine ConvertUnpackOpDynamic, ConvertSignOpDynamic, ConvertSigmoidGradOpDynamic

azazhu commented on tensorflow/compiler/mlir/xla/tests/legalize-tf.mlir:
```diff
@@ -2442,6 +2442,27 @@ func @sigmoid_grad_complex(%arg0: tensor<2xcomplex<f32>>, %arg1: tensor<2xcomple
 
+// CHECK-LABEL: @sigmoid_grad_dynamic
+func @sigmoid_grad_dynamic(%arg0: tensor<?xf32>, %arg1: tensor<?xf32>) -> tensor<?xf32> {
+	// CHECK: shape.shape_of {{.*}} : tensor<?xf32> -> tensor<1xindex>
```
Fixed

azazhu commented on tensorflow/compiler/mlir/xla/transforms/legalize_tf.cc:
```diff
@@ -5088,6 +5088,163 @@ class ConvertUnpackOp : public OpRewritePattern<TF::UnpackOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(TF::UnpackOp op,
+                                     PatternRewriter& rewriter) const override {
```
Fixed.

azazhu commented on tensorflow/compiler/mlir/xla/transforms/legalize_tf.cc:
```diff
@@ -5088,6 +5088,163 @@ class ConvertUnpackOp : public OpRewritePattern<TF::UnpackOp> {
+  LogicalResult matchAndRewrite(TF::UnpackOp op,
+                                     PatternRewriter& rewriter) const override {
+    auto value_type = op.value().getType().cast<RankedTensorType>();
+    if (!value_type || !value_type.hasRank()) return failure();
```
Fixed.

azazhu committed:
add RealDynamicSliceOp's canonicalize and fix CI

azazhu committed:
fix CI for ConvertUnpackOpDynamic

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:
```diff
@@ -1584,6 +1584,75 @@ static LogicalResult Verify(RealDynamicSliceOp op) {
+        limit_op.getValue().dyn_cast_or_null<DenseIntElementsAttr>();
+    auto stride_attr =
+        stride_op.getValue().dyn_cast_or_null<DenseIntElementsAttr>();
+    if (!start_attr || !limit_attr || !stride_attr) failure();
```
You meant to `return failure();` here I think.

(please add a test to cover).

joker-eph replied:
Actually don't bother, I'm fixing it internally

joker-eph replied:
Since there is the issue with the other test and the rebase, please fix this as well! 
Thanks :)

joker-eph commented:
Actually I see a test failure:
```
/build/work/1ba5f054c6ec31340884d6d72bfb5d63b759/google3/runfiles/google3/third_party/tensorflow/compiler/mlir/xla/tests/legalize-tf.mlir:2471:12: error: CHECK: expected string not found in input
 // CHECK: shape.shape_of {{.*}} : tensor<?xf32> -> tensor<1xindex>
           ^
<stdin>:1994:28: note: scanning from here
 func @sigmoid_grad_dynamic(%arg0: tensor<?xf32>, %arg1: tensor<?xf32>) -> tensor<?xf32> {
                           ^
<stdin>:1996:44: note: possible intended match here
 %1 = chlo.broadcast_multiply %arg1, %arg0 : (tensor<?xf32>, tensor<?xf32>) -> tensor<?xf32>
                                           ^
/build/work/1ba5f054c6ec31340884d6d72bfb5d63b759/google3/runfiles/google3/third_party/tensorflow/compiler/mlir/xla/tests/legalize-tf.mlir:2710:12: error: CHECK: expected string not found in input
 // CHECK: "mhlo.sign"(%arg0) : (tensor<?x2x3x?xf32>) -> tensor<?x2x3x?xf32>
           ^
<stdin>:2132:20: note: scanning from here
 %2 = mhlo.constant dense<0.000000e+00> : tensor<f32>
                   ^
<stdin>:2135:56: note: possible intended match here
 %5 = "mhlo.select"(%1, %4, %0) : (tensor<?x2x3x?xi1>, tensor<?x2x3x?xf32>, tensor<?x2x3x?xf32>) -> tensor<?x2x3x?xf32>
                                                       ^

Input file: <stdin>
Check file: /build/work/1ba5f054c6ec31340884d6d72bfb5d63b759/google3/runfiles/google3/third_party/tensorflow/compiler/mlir/xla/tests/legalize-tf.mlir

-dump-input=help explains the following input dump.

Input was:
<<<<<<
              .
              .
              .
           1989:  %1 = mhlo.constant dense<(1.000000e+00,0.000000e+00)> : tensor<2xcomplex<f32>> 
           1990:  %2 = mhlo.subtract %1, %arg0 : tensor<2xcomplex<f32>> 
           1991:  %3 = mhlo.multiply %0, %2 : tensor<2xcomplex<f32>> 
           1992:  return %3 : tensor<2xcomplex<f32>> 
           1993:  } 
           1994:  func @sigmoid_grad_dynamic(%arg0: tensor<?xf32>, %arg1: tensor<?xf32>) -> tensor<?xf32> { 
check:2471'0                                X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           1995:  %0 = mhlo.constant dense<1.000000e+00> : tensor<f32> 
check:2471'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           1996:  %1 = chlo.broadcast_multiply %arg1, %arg0 : (tensor<?xf32>, tensor<?xf32>) -> tensor<?xf32> 
check:2471'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:2471'1                                                ?                                                  possible intended match
           1997:  %2 = chlo.broadcast_subtract %0, %arg0 {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<f32>, tensor<?xf32>) -> tensor<?xf32> 
check:2471'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           1998:  %3 = chlo.broadcast_multiply %1, %2 : (tensor<?xf32>, tensor<?xf32>) -> tensor<?xf32> 
check:2471'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           1999:  return %3 : tensor<?xf32> 
check:2471'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           2000:  } 
check:2471'0     ~~~
           2001:  func @sin(%arg0: tensor<2xf32>) -> tensor<2xf32> { 
check:2471'0     ~~~~~~~~~~
              .
              .
              .
           2127:  return %0 : tensor<1x2x3x4xf32> 
           2128:  } 
           2129:  func @sign_dynamic(%arg0: tensor<?x2x3x?xf32>) -> tensor<?x2x3x?xf32> { 
           2130:  %0 = "mhlo.sign"(%arg0) : (tensor<?x2x3x?xf32>) -> tensor<?x2x3x?xf32> 
           2131:  %1 = "mhlo.compare"(%arg0, %arg0) {comparison_direction = "NE"} : (tensor<?x2x3x?xf32>, tensor<?x2x3x?xf32>) -> tensor<?x2x3x?xi1> 
           2132:  %2 = mhlo.constant dense<0.000000e+00> : tensor<f32> 
check:2710'0                        X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           2133:  %3 = shape.shape_of %arg0 : tensor<?x2x3x?xf32> -> tensor<4xindex> 
check:2710'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           2134:  %4 = "mhlo.dynamic_broadcast_in_dim"(%2, %3) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<f32>, tensor<4xindex>) -> tensor<?x2x3x?xf32> 
check:2710'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           2135:  %5 = "mhlo.select"(%1, %4, %0) : (tensor<?x2x3x?xi1>, tensor<?x2x3x?xf32>, tensor<?x2x3x?xf32>) -> tensor<?x2x3x?xf32> 
check:2710'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:2710'1                                                            ?                                                                 possible intended match
           2136:  return %5 : tensor<?x2x3x?xf32> 
check:2710'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           2137:  } 
check:2710'0     ~~~
           2138:  func @slice_constant_start(%arg0: tensor<4xi32>) -> tensor<2xi32> { 
check:2710'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           2139:  %0 = mhlo.constant dense<1> : tensor<1xi64> 
           2140:  %1 = mhlo.constant dense<2> : tensor<1xi64> 
              .
              .
              .
```

I haven't debugged it, if you can take a look that'd be great!

azazhu committed:
rebase

azazhu committed:
fix typo in ConvertSigmoidGradOpDynamic

azazhu commented:
> Actually I see a test failure:
> 
> ```
> /build/work/1ba5f054c6ec31340884d6d72bfb5d63b759/google3/runfiles/google3/third_party/tensorflow/compiler/mlir/xla/tests/legalize-tf.mlir:2471:12: error: CHECK: expected string not found in input
>  // CHECK: shape.shape_of {{.*}} : tensor<?xf32> -> tensor<1xindex>
>            ^
> <stdin>:1994:28: note: scanning from here
>  func @sigmoid_grad_dynamic(%arg0: tensor<?xf32>, %arg1: tensor<?xf32>) -> tensor<?xf32> {
>                            ^
> <stdin>:1996:44: note: possible intended match here
>  %1 = chlo.broadcast_multiply %arg1, %arg0 : (tensor<?xf32>, tensor<?xf32>) -> tensor<?xf32>
>                                            ^
> /build/work/1ba5f054c6ec31340884d6d72bfb5d63b759/google3/runfiles/google3/third_party/tensorflow/compiler/mlir/xla/tests/legalize-tf.mlir:2710:12: error: CHECK: expected string not found in input
>  // CHECK: "mhlo.sign"(%arg0) : (tensor<?x2x3x?xf32>) -> tensor<?x2x3x?xf32>
>            ^
> <stdin>:2132:20: note: scanning from here
>  %2 = mhlo.constant dense<0.000000e+00> : tensor<f32>
>                    ^
> <stdin>:2135:56: note: possible intended match here
>  %5 = "mhlo.select"(%1, %4, %0) : (tensor<?x2x3x?xi1>, tensor<?x2x3x?xf32>, tensor<?x2x3x?xf32>) -> tensor<?x2x3x?xf32>
>                                                        ^
> 
> Input file: <stdin>
> Check file: /build/work/1ba5f054c6ec31340884d6d72bfb5d63b759/google3/runfiles/google3/third_party/tensorflow/compiler/mlir/xla/tests/legalize-tf.mlir
> 
> -dump-input=help explains the following input dump.
> 
> Input was:
> <<<<<<
>               .
>               .
>               .
>            1989:  %1 = mhlo.constant dense<(1.000000e+00,0.000000e+00)> : tensor<2xcomplex<f32>> 
>            1990:  %2 = mhlo.subtract %1, %arg0 : tensor<2xcomplex<f32>> 
>            1991:  %3 = mhlo.multiply %0, %2 : tensor<2xcomplex<f32>> 
>            1992:  return %3 : tensor<2xcomplex<f32>> 
>            1993:  } 
>            1994:  func @sigmoid_grad_dynamic(%arg0: tensor<?xf32>, %arg1: tensor<?xf32>) -> tensor<?xf32> { 
> check:2471'0                                X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
>            1995:  %0 = mhlo.constant dense<1.000000e+00> : tensor<f32> 
> check:2471'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            1996:  %1 = chlo.broadcast_multiply %arg1, %arg0 : (tensor<?xf32>, tensor<?xf32>) -> tensor<?xf32> 
> check:2471'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> check:2471'1                                                ?                                                  possible intended match
>            1997:  %2 = chlo.broadcast_subtract %0, %arg0 {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<f32>, tensor<?xf32>) -> tensor<?xf32> 
> check:2471'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            1998:  %3 = chlo.broadcast_multiply %1, %2 : (tensor<?xf32>, tensor<?xf32>) -> tensor<?xf32> 
> check:2471'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            1999:  return %3 : tensor<?xf32> 
> check:2471'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            2000:  } 
> check:2471'0     ~~~
>            2001:  func @sin(%arg0: tensor<2xf32>) -> tensor<2xf32> { 
> check:2471'0     ~~~~~~~~~~
>               .
>               .
>               .
>            2127:  return %0 : tensor<1x2x3x4xf32> 
>            2128:  } 
>            2129:  func @sign_dynamic(%arg0: tensor<?x2x3x?xf32>) -> tensor<?x2x3x?xf32> { 
>            2130:  %0 = "mhlo.sign"(%arg0) : (tensor<?x2x3x?xf32>) -> tensor<?x2x3x?xf32> 
>            2131:  %1 = "mhlo.compare"(%arg0, %arg0) {comparison_direction = "NE"} : (tensor<?x2x3x?xf32>, tensor<?x2x3x?xf32>) -> tensor<?x2x3x?xi1> 
>            2132:  %2 = mhlo.constant dense<0.000000e+00> : tensor<f32> 
> check:2710'0                        X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
>            2133:  %3 = shape.shape_of %arg0 : tensor<?x2x3x?xf32> -> tensor<4xindex> 
> check:2710'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            2134:  %4 = "mhlo.dynamic_broadcast_in_dim"(%2, %3) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<f32>, tensor<4xindex>) -> tensor<?x2x3x?xf32> 
> check:2710'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            2135:  %5 = "mhlo.select"(%1, %4, %0) : (tensor<?x2x3x?xi1>, tensor<?x2x3x?xf32>, tensor<?x2x3x?xf32>) -> tensor<?x2x3x?xf32> 
> check:2710'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> check:2710'1                                                            ?                                                                 possible intended match
>            2136:  return %5 : tensor<?x2x3x?xf32> 
> check:2710'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            2137:  } 
> check:2710'0     ~~~
>            2138:  func @slice_constant_start(%arg0: tensor<4xi32>) -> tensor<2xi32> { 
> check:2710'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            2139:  %0 = mhlo.constant dense<1> : tensor<1xi64> 
>            2140:  %1 = mhlo.constant dense<2> : tensor<1xi64> 
>               .
>               .
>               .
> ```
> 
> I haven't debugged it, if you can take a look that'd be great! (and rebase as well)

Mehdi, I rebased and fixed this failure.(The root cause is typo "_return_  failure()")

joker-eph commented:
> Mehdi, I rebased and fixed this failure.(The root cause is typo "return failure()")

It seems to fail the same way right now?

azazhu committed:
fix tests: not extend chlo

azazhu commented:
> > Mehdi, I rebased and fixed this failure.(The root cause is typo "return failure()")
> 
> It seems to still fail the same way at the moment?

Sorry, I extended chlo. Fixed.

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:
```diff
@@ -1626,6 +1626,75 @@ static LogicalResult Verify(RealDynamicSliceOp op) {
+  using OpRewritePattern<RealDynamicSliceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(RealDynamicSliceOp real_dynamic_slice,
+                                PatternRewriter& rewriter) const override {
```
It seems that this pattern isn't invoked right now (according to our test coverage), are you missing tests?

rthadur commented:
Seems auto-merge is not happening but the changes are now committed so we can close this. Thank you for the PR.
